temporal-sdk/src/main/java/io/temporal/worker/tuning/SlotSupplier.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (C) 2022 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ * Copyright (C) 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this material except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.worker.tuning;
+
+import io.temporal.common.Experimental;
+import java.util.Optional;
+
+/**
+ * A SlotSupplier is responsible for managing the number of slots available for a given type of
+ * task. The three types of tasks are workflow, activity, and local activity. Implementing this
+ * interface allows you to carefully control how many tasks of any given type a worker will process
+ * at once.
+ *
+ * @param <SI> The type of information that will be used to reserve a slot. The three info types are
+ *     {@link WorkflowSlotInfo}, {@link ActivitySlotInfo}, and {@link LocalActivitySlotInfo}.
+ */
+@Experimental
+public interface SlotSupplier<SI> {
+  /**
+   * This function is called before polling for new tasks. Your implementation should block until a
+   * slot is available, then return a permit to use that slot.
+   *
+   * @param ctx The context for slot reservation.
+   * @return A permit to use the slot which may be populated with your own data.
+   * @throws InterruptedException The worker may choose to interrupt the thread in order to cancel
+   *     the reservation, or during shutdown. You may perform cleanup, and then should rethrow the
+   *     exception.
+   */
+  SlotPermit reserveSlot(SlotReservationContext<SI> ctx) throws InterruptedException;
```

Comment:
I like having the internal ID because it ensures equality & hashing always work properly, regardless of what the user ends up doing with their data (yeah, objects have an ID, but they could end up implementing eq/hash at the top level in a way that potentially breaks things - but to your point we could get rid of the internal ID and still have that work seemingly), and it also makes for a nice, clear type at the top-level. The type name alone feels worth it to me.

Reply:
> yeah, objects have an ID, but they could end up implementing eq/hash at the top level in a way that potentially breaks things - but to your point we could get rid of the internal ID and still have that work seemingly

I'd have to understand this a bit better. Equals/hashCode will work for `new Object()` directly and if someone overrides those for a specific object, I would expect them to work too. I don't think we should ever invoke equals/hashCode on this object for anything though.

Reply:
The hash/eq is used because we use it in the hash map in the tracking supplier. So, sure, we expect it to work just fine - but our implementation will always be correct (and likely cheaper) than a user-supplied one.

Reply:
I think you should then make `SlotPermit` `final` (so nobody extends it and messes up this expectation) and you should rely on the default equals/hashCode which is identity based (i.e. don't override it).